(feat) i18n - Added support to translate error messages and other form fields

icrc-jofrancisco opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.

## Summary

This pull request adds support for translating error messages and labels within a form. Labels are now provided by a JSON file, following the naming convention "<lang>.json". This update provides a more user-friendly experience for non-native speakers.

Currently, the form supports English and French languages, and in the future, additional languages can be added by creating a new JSON file with the appropriate naming convention.



In addition to updating the form to display translated labels, this pull request also updates some core components, including buttons and error messages, to be translatable. The updated tests ensure that the translation module is properly imported and functioning as expected, improving the overall quality and reliability of the codebase.
<!--
Required.
Please describe what problems your PR addresses.
-->


## Screenshots

![Pasted Graphic 3](https://user-images.githubusercontent.com/94977371/225369847-17085841-56e3-449f-b67d-4ba56534a860.png)


Thanks,

ibacher commented:
Thanks @icrc-jofrancisco! This looks extremely useful; however, from a maintainability perspective, it would be better if we could leverage i18next, e.g., by using [this library](https://github.com/Romanchuk/angular-i18next).

For similar reasons, it would be good if the translation keys are in keeping with the keys we use in other repos, i.e., `enterMoreCharacters` rather than `ENTER_MORE_CHARACTERS`.

Also, looking through this quickly, I notice that your included translations are part of the test app itself rather than the engine, so I'm really unclear how this would actually work at runtime.

Finally, have you tested this with the translation feature already merged into the engine? I'm worried that the two of these would wind up being incompatible.

icrc-jofrancisco commented:
> Thanks @icrc-jofrancisco! This looks extremely useful; however, from a maintainability perspective, it would be better if we could leverage i18next, e.g., by using [this library](https://github.com/Romanchuk/angular-i18next).
> 
> For similar reasons, it would be good if the translation keys are in keeping with the keys we use in other repos, i.e., `enterMoreCharacters` rather than `ENTER_MORE_CHARACTERS`.
> 
> Also, looking through this quickly, I notice that your included translations are part of the test app itself rather than the engine, so I'm really unclear how this would actually work at runtime.
> 
> Finally, have you tested this with the translation feature already merged into the engine? I'm worried that the two of these would wind up being incompatible.

Hi @ibacher 
Thank you for your message and suggestions. I understand your concerns regarding maintainability and consistency with the translation keys used in other repos.

Regarding the library used for translation, I am currently using `ngx-translate ` from Angular because it is already being used in the project for translating concepts.

Regarding the labels being part of the test app, I can confirm that the module is initialized there along with other labels. However, after testing and confirmation, I plan to create a PR on the patient chart to create the labels. With this approach, we can have control of the labels in the patient chart without touching the engine itself.

Regarding the naming convention of the translation keys, I followed the same nomenclature as before in the messages.ts file. However, I appreciate your suggestion and am open to following the suggested convention of using “enterMoreCharacters” instead of “ENTER_MORE_CHARACTERS” for better consistency across repos.

Thank you for your feedback!

ibacher commented:
Sorry, I have updated this somewhat, but I'd hoped in non-destructive ways, and mostly to fix conflicts with one of Andre's PRs.

icrc-jofrancisco commented:
> Sorry, I have updated this somewhat, but I'd hoped in non-destructive ways, and mostly to fix conflicts with one of Andre's PRs.



> So, when I run this using the corresponding PR branch in the chart, but without supplying any additional translations, I get this:
> 
> ![Screenshot showing REQUIRED_FIELD_MSG](https://user-images.githubusercontent.com/52504170/246857853-bb3ec69b-b687-4ab9-b28a-2bcf32996986.png)
> 
> Any idea what I missed in doing this?
> 
> Steps to reproduce:
> 
> 1. Checkout this branch
> 2. Run `yarn build:lib`
> 3. Run `touch dist/ngx-formentry/yarn.lock`
> 4. In Chart, link to directory containing dist
> 5. Run Chart using `yarn start --sources 'packages/esm-patient-form-app' --sources 'packages/esm-form-entry-app'`

Hey @ibacher , thanks a lot for checking out my pull request! I wanted to give you a heads-up that I already made some updates based on your feedback. 

I noticed a couple of translation keys in the ngx-formentry module that were off (because it was decided to change the numenclature and there were 2 wrong keys, my bad..).  
I've fixed them now. 

I would appreciate it if you could continue the review, I am open to further suggestions.
Thanks,
